chain/src/sumtree.rs

```diff
@@ -89,27 +88,74 @@ impl SumTrees {
 		})
 	}
 
-	/// Whether a given commitment exists in the Output MMR and it's unspent
-	pub fn is_unspent(&mut self, commit: &Commitment) -> Result<bool, Error> {
-		let rpos = self.commit_index.get_output_pos(commit);
-		match rpos {
+	pub fn is_unspent(&mut self, output: &OutputIdentifier) -> Result<(), Error> {
+		match self.commit_index.get_output_pos(&output.commit) {
+			Ok(pos) => {
+				let output_pmmr = PMMR::at(
+					&mut self.output_pmmr_h.backend,
+					self.output_pmmr_h.last_pos,
+				);
+				if let Some(HashSum { hash, sum: _ }) = output_pmmr.get(pos) {
+					let sum_commit = output.as_sum_commit();
+					let hash_sum = HashSum::from_summable(pos, &sum_commit);
+					if hash == hash_sum.hash {
+						Ok(())
+					} else {
+						Err(Error::SumTreeErr(format!("sumtree hash mismatch")))
+					}
+				} else {
+					Err(Error::OutputNotFound)
+				}
+			}
+			Err(grin_store::Error::NotFoundErr) => Err(Error::OutputNotFound),
+			Err(e) => Err(Error::StoreErr(e, format!("sumtree unspent check"))),
+		}
+	}
+
+	pub fn is_matured(
+		&mut self,
+		input: &Input,
+		height: u64,
+	) -> Result<(), Error> {
+		match self.commit_index.get_output_pos(&input.commit) {
 			Ok(pos) => {
 				let output_pmmr = PMMR::at(
 					&mut self.output_pmmr_h.backend,
-					self.output_pmmr_h.last_pos
+					self.output_pmmr_h.last_pos,
 				);
-				if let Some(hs) = output_pmmr.get(pos) {
-					let hashsum = HashSum::from_summable(
-						pos,
-						&SumCommit::from_commit(&commit),
-					);
-					Ok(hs.hash == hashsum.hash)
+				if let Some(HashSum { hash, sum: _ }) = output_pmmr.get(pos) {
+					let sum_commit = SumCommit::from_input(&input);
+					let hash_sum = HashSum::from_summable(pos, &sum_commit);
+					if hash != hash_sum.hash {
+						return Err(Error::SumTreeErr(format!("sumtree hash mismatch")));
+					}
+
+					// At this point we can be sure the input is spending the output
+					// it claims to be spending, and it is coinbase or non-coinbase.
+					// If we are spending a coinbase output then go find the block
+					// and check the coinbase maturity rule is being met.
+					if input.features.contains(COINBASE_OUTPUT) {
+						// edge case here - we may be spending an output with zero-confirmations
+						// and it just got included in a block.
+						// The wallet may not know about this new block yet (not yet refreshed).
+						// Is it safe to assume this the tx was included in the latest block?
```

Comment:
If I understand well, I don't think we need to worry about that. We explicitly disallow inputs spending outputs in the same block, as they should have been cut-through. This happens in input validation as we're doing it before doing the outputs.

Reply:
Yeah - good catch. Just revisiting this now and this is unnecessary.